### PR TITLE
Publish Gem from Jenkins

### PR DESCRIPTION
**Pull request by alecgibson**
This moves the logic for Gem publishing away from the [Gem Publisher](https://github.com/alphagov/gem_publisher) gem, and into the `Jenkinsfile`, where it is executed as part of the build chain instead.

The motivation for this change is that getting Gem Publisher to work with the new Jenkins 2 CI environment has been non-trivial (and remains unsolved on eg [API Adapters](https://github.com/alphagov/gds-api-adapters)).

If this is a successful and accepted approach, this could be extracted into Puppet's [`govuk_jenkinslib.groovy`](https://github.com/alphagov/govuk-puppet/blob/master/modules/govuk_jenkins/files/var/lib/jenkins/groovy_scripts/govuk_jenkinslib.groovy) file for use across projects.

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -33,9 +33,29 @@ node {
...
-        sshagent(['govuk-ci-ssh-key']) {
-          govuk.runRakeTask("publish_gem --trace")
+        def version = sh(
+          script: /ruby -e "puts eval(File.read('govuk_navigation_helpers.gemspec'), TOPLEVEL_BINDING).version.to_s"/,
+          returnStdout: true
+        ).trim()
+
+        echo 'This version: ' + version
+
+        def shouldPush = sh(
```
Comment (boffbowsh) [inline Jenkinsfile]: We should also check if the gem is pushed in rubygems:
```bash
~$ gem list -rq gds-api-adapters | grep '(39.2.0)' ; echo $?
gds-api-adapters (39.2.0)
0
~$ gem list -rq gds-api-adapters | grep '(39.3.0)' ; echo $?
1
```

Comment (alecgibson) [inline Jenkinsfile]: Makes sense to me - does that mean we're going to try and go ahead with this approach?

Comment (boffbowsh) [inline Jenkinsfile]: Yes. I think we use this gem as a canary for this method, and if it works well move the logic into the `govuk_jenkinslib` for other gems to use

Comment (alecgibson) [inline Jenkinsfile]: Will we eventually look into deprecating the Gem Publisher gem? Or will some build chains still require it? Basically, is it weird to have two ways to publish? (I personally think the logic is simple enough that this isn't a big deal)

Comment (boffbowsh) [inline Jenkinsfile]: Adding the shared logic into Jenkins would effectively deprecate that gem.

File: Jenkinsfile [hunk 2]
```diff
@@ -33,9 +33,29 @@ node {
 
     if(env.BRANCH_NAME == "master") {
       stage('Publish Gem') {
-        sshagent(['govuk-ci-ssh-key']) {
-          govuk.runRakeTask("publish_gem --trace")
+        def version = sh(
+          script: /ruby -e "puts eval(File.read('govuk_navigation_helpers.gemspec'), TOPLEVEL_BINDING).version.to_s"/,
```
Comment (alecgibson) [inline Jenkinsfile]: Does anyone have any bright ideas on this? I think it's a bit weird to be running Ruby inside the shell inside Groovy (there's just so many layers it's kind of painful), but I can't think of a better way of getting at the version number.

File: Jenkinsfile (hunk 1 above)
Comment (alecgibson) [inline Jenkinsfile]: @boffbowsh I've pushed that change

File: Jenkinsfile [hunk 3]
```diff
@@ -33,8 +35,34 @@ node {
...
+
+        if (taggedReleaseExists) {
+          echo "Version ${version} has already been tagged on Github"
+        } else {
+          echo('Pushing tag')
+          govuk.pushTag(REPOSITORY, env.BRANCH_NAME, 'v' + version)
+        }
+
+        def versionAlreadyPublished = sh(
+          script: "gem list ^${REPOSITORY}\$ --remote --all --quiet | grep ${version}",
```
Comment (boffbowsh) [inline Jenkinsfile]: Might be good to anchor the grep within parentheses, so `grep 1.2.3` doesn't match `11.2.3` for example. So `grep '(${version})'`

Comment (alecgibson) [inline Jenkinsfile]: Oh, do you mean drop the `--all` flag and assume we should only be checking the latest version?

Comment (boffbowsh) [inline Jenkinsfile]: No, that's fine, I mean that the `grep` isn't anchored well enough, it's possible for it to match things it shouldn't do.

Comment (alecgibson) [inline Jenkinsfile]: @boffbowsh I've fixed the regex.

File: Jenkinsfile [hunk 4]
```diff
@@ -31,13 +33,40 @@ node {
       govuk.runTests('spec')
     }
 
-    if(env.BRANCH_NAME == "master") {
+    //if(env.BRANCH_NAME == "master") {
```
Comment (boffbowsh) [inline Jenkinsfile]: I'm guessing you don't want this commented out?

Comment (alecgibson) [inline Jenkinsfile]: Oh yeah whoops - commented for testing. Will remove.

**Comment by alecgibson** [comment]
@tomnatt any thoughts on this?

**Pull request merged**